dev/Core#53 Add in stats to AB tests and add in clickthrough and open…

seamuslee001 opened the pull request:
…ed % stats to mailing reports

Overview
----------------------------------------
This adds Mailing delivery stat percentages to AB tests and also adds in new opened and clickthrough rate stats

Before
----------------------------------------
No AB testing % stats
![screenshot 2018-04-09 14 20 48](https://user-images.githubusercontent.com/336308/38476387-93dc6eea-3c01-11e8-82bc-0142894b5502.png)


After
----------------------------------------
AB Tests have % stats and rates. 

![screenshot 2018-04-09 14 23 05](https://user-images.githubusercontent.com/336308/38476393-99045da6-3c01-11e8-8159-6c90db2c23ce.png)

eileenmcnaughton commented:
test this please

eileenmcnaughton commented:
@seamuslee001 will trade for #11954

seamuslee001 commented:
Deal @eileenmcnaughton

On Mon, 9 Apr 2018 at 9:36 am, Eileen McNaughton <notifications@github.com>
wrote:

> @seamuslee001 <https://github.com/seamuslee001> will trade for #11954
> <https://github.com/civicrm/civicrm-core/pull/11954>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/civicrm/civicrm-core/pull/11957#issuecomment-379591872>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGe_FS0z75d1F-Ul6USs9waJUT9u_TFOks5tmp7-gaJpZM4TLR17>
> .
>

eileenmcnaughton commented on CRM/Mailing/BAO/Mailing.php:
```diff
@@ -1997,12 +1997,16 @@ public static function &report($id, $skipDetails = FALSE, $isSMS = FALSE) {
         $row['bounce_rate'] = (100.0 * $mailing->bounce) / $mailing->queue;
         $row['unsubscribe_rate'] = (100.0 * $row['unsubscribe']) / $mailing->queue;
         $row['optout_rate'] = (100.0 * $row['optout']) / $mailing->queue;
+        $row['opened_rate'] = ($row['opened'] / $mailing->delivered * 100.0);
```
We should probably add protection against a hard error if delivered === 0 (presumably a very obscure situation but worth handling I think)

seamuslee001 replied:
yeh i guess so

eileenmcnaughton replied:
Also - can we stdise bracketting - I prefer but mostly just thing same is better 
```
 $row['opened_rate'] = ($row['opened'] / $mailing->delivered) * 100;
```
with conditional

 $row['opened_rate'] = $mailing->delivered ?  (($row['opened'] / $mailing->delivered) * 100) : '';

eileenmcnaughton commented:
@seamuslee001 I agree with this change in principle - ie it adds additional useful information to the A-B stats report and the change is localised to that report & hence low risk. I don't believe it could have a negative impact if it works as intended.

Per comments I think each divided by line has a risk of being divided by 0 so worth going through those - in fact my test did hit that on the second mailing (I added the screen shots to the PR description above - but perhaps the after might get replaced again before merge).

Visually I think

Tracked Opens 1 (100%) is better than Tracked Opens 1 100%

I think I'd also prefer the link to be on the number & not the % but from a quick look at the code that looks like more hassle than it's worth to do that.

seamuslee001 commented:
@eileenmcnaughton thanks for that I think i have now addressed all issues you alluded to in your review. I haven't quite got the () around on AB tests but there is at least now a ui difference between the number and the %

eileenmcnaughton commented on api/v3/Mailing.php:
```diff
@@ -769,6 +769,9 @@ function civicrm_api3_mailing_stats($params) {
         break;
     }
   }
+  $stats[$params['mailing_id']]['delivered_rate'] = round((100.0 * $stats[$params['mailing_id']]['Delivered']) / CRM_Mailing_Event_BAO_Queue::getTotalCount($params['mailing_id'], $params['job_id']), 2) . '%';
```
could these be subject to division by zero?

eileenmcnaughton commented:
@seamuslee001 I'm still seeing my Granny on the B side
![screenshot 2018-04-09 16 11 34](https://user-images.githubusercontent.com/336308/38479021-ba492ec4-3c10-11e8-95c8-05552433087c.png)

eileenmcnaughton commented:
Yes - there is an inconsistency in 0 vs na but that is not new in this PR. The grey power is gone!

![screenshot 2018-04-09 16 25 48](https://user-images.githubusercontent.com/336308/38479358-b8215340-3c12-11e8-9d24-8e5c2a22bedf.png)

eileenmcnaughton commented:
I'm happy this is ready to merge now - the things I observed have been addressed

seamuslee001 commented:
Merging as it indicates merge-on-pass
